Mathlib/AlgebraicGeometry/Morphisms/RingHomProperties.lean

```diff
@@ -38,504 +61,311 @@ universe u
+    Scheme.ΓSpecIso_naturality, H.cancel_left_isIso]
+
+theorem of_iSup_eq_top [IsAffine Y] {ι : Type*}
+    (U : ι → X.affineOpens) (hU : ⨆ i, (U i : Opens X) = ⊤)
+    (H : ∀ i, Q (f.appLE ⊤ (U i).1 le_top)) :
+    P f := by
+  have (i) : IsAffine ((X.openCoverOfSuprEqTop _ hU).obj i) := (U i).2
+  refine of_openCover (X.openCoverOfSuprEqTop _ hU) fun i ↦ ?_
+  simpa [Scheme.Hom.app_eq_appLE] using (f.appLE_congr _ rfl (by simp) Q).mp (H i)
+
+theorem iff_of_iSup_eq_top [IsAffine Y] {ι : Type*}
+    (U : ι → X.affineOpens) (hU : ⨆ i, (U i : Opens X) = ⊤) :
+    P f ↔ ∀ i, Q (f.appLE ⊤ (U i).1 le_top) :=
+  ⟨fun H _ ↦ appLE P f H ⟨_, isAffineOpen_top _⟩ _ le_top, of_iSup_eq_top U hU⟩
+
+lemma HasAffineProperty.isLocalAtSource (P) {Q} [HasAffineProperty P Q]
+    (H : ∀ {X Y : Scheme.{u}} (f : X ⟶ Y) [IsAffine Y] (𝒰 : Scheme.OpenCover.{u} X),
+        Q f ↔ ∀ i, Q (𝒰.map i ≫ f)) : IsLocalAtSource P where
+  iff_of_openCover' {X Y} f 𝒰 := by
+    simp_rw [IsLocalAtTarget.iff_of_iSup_eq_top _ (iSup_affineOpens_eq_top Y)]
+    rw [forall_comm]
+    refine forall_congr' fun U ↦ ?_
+    simp_rw [HasAffineProperty.iff_of_isAffine, morphismRestrict_comp]
+    exact @H _ _ (f ∣_ U.1) U.2 (𝒰.restrict (f ⁻¹ᵁ U.1))
+
+instance : IsLocalAtSource P := by
+  apply HasAffineProperty.isLocalAtSource
+  intros X Y f _ 𝒰
+  simp_rw [← HasAffineProperty.iff_of_isAffine (P := P),
+    iff_of_openCover 𝒰.affineRefinement.openCover,
+    fun i ↦ iff_of_openCover (P := P) (f := 𝒰.map i ≫ f) (𝒰.obj i).affineCover]
+  simp [Scheme.OpenCover.affineRefinement]
+
+instance : P.ContainsIdentities where
+  id_mem X := by
+    rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := P) _ (iSup_affineOpens_eq_top _)]
+    intro U
+    rw [morphismRestrict_id, iff_of_isAffine (P := P), Scheme.id_app]
+    have := IsLocalization.at_units (.powers (1 : Γ(X ∣_ᵤ 𝟙 X ⁻¹ᵁ U, ⊤))) (by simp)
+    exact (isLocal_ringHomProperty P).HoldsForLocalizationAway Γ(X ∣_ᵤ 𝟙 X ⁻¹ᵁ U, ⊤) 1
+
+instance : P.IsStableUnderComposition where
+  comp_mem {X Y Z} f g hf hg := by
+    wlog hZ : IsAffine Z generalizing X Y Z
+    · rw [IsLocalAtTarget.iff_of_iSup_eq_top (P := P) _ (iSup_affineOpens_eq_top _)]
+      intro U
+      rw [morphismRestrict_comp]
+      exact this _ _ (IsLocalAtTarget.restrict hf _) (IsLocalAtTarget.restrict hg _) U.2
+    wlog hY : IsAffine Y generalizing X Y
+    · rw [IsLocalAtSource.iff_of_openCover (P := P) (Y.affineCover.pullbackCover f)]
+      intro i
+      rw [← Scheme.OpenCover.pullbackHom_map_assoc]
+      exact this _ _ (IsLocalAtTarget.of_isPullback (.of_hasPullback _ _) hf)
+        (comp_of_isOpenImmersion _ _ _ hg) inferInstance
+    wlog hX : IsAffine X generalizing X
+    · rw [IsLocalAtSource.iff_of_openCover (P := P) X.affineCover]
+      intro i
+      rw [← Category.assoc]
+      exact this _ (comp_of_isOpenImmersion _ _ _ hf) inferInstance
+    rw [iff_of_isAffine (P := P)] at hf hg ⊢
+    exact (isLocal_ringHomProperty P).StableUnderComposition _ _ hg hf
```

Comment:
Nice!

Reply:
This code looks better but ideally we should have some automation to do it for us...